python/ray/tune/utils/util.py

```diff
@@ -124,18 +125,31 @@ def stop(self):
 @DeveloperAPI
 def retry_fn(
     fn: Callable[[], Any],
-    exception_type: Type[Exception],
+    exception_type: Type[Exception] = Exception,
     num_retries: int = 3,
     sleep_time: int = 1,
-):
-    for i in range(num_retries):
-        try:
-            fn()
-        except exception_type as e:
-            logger.warning(e)
-            time.sleep(sleep_time)
-        else:
-            break
+    timeout: Optional[Number] = None,
+) -> bool:
+    def _retry_fn():
+        for i in range(num_retries):
+            try:
+                fn()
+            except exception_type as e:
+                logger.warning(e)
+                time.sleep(sleep_time)
+            else:
+                return
+
+    proc = threading.Thread(target=_retry_fn)
+    proc.daemon = True
+    proc.start()
+    proc.join(timeout=timeout)
```

Comment:
now that you have a thread, imagine eventually we checkpoint on the side while the training just keeps going 🤯 😄 

one nit, I also think timeout should be per-retry? (so ```timeout=num_retries * timeout``` here). otherwise the actual timeout will be dependent on how many retries you set here? although, admittedly, num_retries is not even a configurable bit.

Reply:
Yeah I thought about this, but the reason why I kept a global timeout was because it is a) simpler/cleaner to implement and b) we basically want to define a maximum time we want to block training, so I think we should be fine with this. Let me know if you prefer this per-retry

Reply:
Ah actually as discussed, let's have a timeout per retry. Otherwise if the first sync hangs we will not try again. Updated the PR